plugin.video.retrospect/resources/libs/helpers/channelimporter.py

```diff
@@ -0,0 +1,707 @@
+        else:
+            addon_path = os.path.abspath(os.path.join(Config.rootDir, ".."))
+
+        return addon_path
+
+    def __is_channel_set_updated(self, channel_info):
+        """ Checks whether a channel set was updated.
+
+        :param ChannelInfo channel_info: the channelInfo for a channel from the set
+
+        :return: indicating if the channel was updated or not.
+        :rtype : bool
+
+        """
+
+        compiled_name = "%s.pyc" % (channel_info.moduleName,)
+        optimized_name = "%s.pyo" % (channel_info.moduleName,)
+
+        # A channel set is updated if no Optimized (.pyo) and no Compiled (.pyc) files are
+        # there.
+        if os.path.isfile(os.path.join(channel_info.path, compiled_name)) or \
+                os.path.isfile(os.path.join(channel_info.path, optimized_name)):
+            return False
+
+        # If we run Python 3 and no files are in the Python3 __pycache__ folders a channel is
+        # also considered updated.
+        if PY3 and glob.glob(os.path.join(channel_info.path, "__pycache__", "*.py*")):
+            return False
+
+        return True
+
+    def __channel_is_correct(self, channel_info):
+        # type: (ChannelInfo) -> bool
+        """ Validates if the given channel with channelInfo is correct
+
+        @param channel_info: The channelInfo to use to validate the channel
+        @return:            True/False if valid or not.
+
+        """
+
+        if not channel_info.guid:
+            Logger.error("Not loading: %s -> No guid present.", channel_info)
+            return False
+
+        if channel_info in self.__allChannels:
+            existing_channel = self.__allChannels[self.__allChannels.index(channel_info)]
+            Logger.error("Not loading: %s -> a channel with the same guid already exist:\n%s.",
+                         channel_info, existing_channel)
+            return False
+
+        return True
+
+    def __initialise_channel_set(self, channel_info):
+        # type: (ChannelInfo) -> None
+        """ Initialises a channelset (.py file)
+
+        WARNING: these actions are done ONCE per python file, not per channel.
+
+        Arguments:
+        channelInfo : ChannelInfo - The channelinfo
+
+        Keyword Arguments:
+        abortOnNew  : Boolean - If set to true, channel initialisation will not continue if a new channel was found.
+                                This will have to be done later.
+
+        Returns True if any operations where executed
+
+        """
+
+        Logger.info("Initialising channel set at: %s.", channel_info.path)
+
+        # now import (required for the PerformFirstTimeActions
+        sys.path.append(channel_info.path)
+
+        # make sure a pyo or pyc exists
+        # __import__(channelInfo.moduleName)
+        # The debugger won't compile if __import__ is used. So let's use this one.
+        import py_compile
```

Comment:
why? python interpreter should do this automatically.

Reply:
I does, but for Retrospect, a missing pyo/pyc file is an indication that there was never encountered before (after add-on installation or updating) and this triggers some initialisations, such as showing a first time message. This can however happen without actually importing the corresponding Python file and hence it would not be compiled. This would trigger a second (and unwanted) initialisation if Retrospect runs again because it will again find the file without pyo/pyc. That is why I manually compile after an add-on update.

Keep in mind, that this happens on the user's machine. I don't include precompiled Python files.